server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/remote/delivery/BouncerTest.java

```diff
@@ -417,4 +422,34 @@ public void bounceShouldWorkWhenProcessorSpecifiedAndNoExceptionMessage() throws
         assertThat(mailetContext.getSentMails()).containsOnly(expected);
         assertThat(mailetContext.getBouncedMails()).isEmpty();
     }
+
+    @Test
+    public void bounceShouldAttachErrorCodeWhenSmtpError() throws Exception {
+        RemoteDeliveryConfiguration configuration = new RemoteDeliveryConfiguration(
+            FakeMailetConfig.builder()
+                .setProperty(RemoteDeliveryConfiguration.HELO_NAME, HELLO_NAME)
+                .setProperty(RemoteDeliveryConfiguration.BOUNCE_PROCESSOR, BOUNCE_PROCESSOR)
+                .build(),
+            mock(DomainList.class));
+        Bouncer testee = new Bouncer(configuration, mailetContext);
+
+        Mail mail = FakeMail.builder().name("name").state(Mail.DEFAULT)
+            .sender(MailAddressFixture.ANY_AT_JAMES)
+            .build();
+
+        SMTPSendFailedException ex = mock(SMTPSendFailedException.class);
+        when(ex.getReturnCode()).thenReturn(SMTP_ERROR_CODE_521);
+
+        testee.bounce(mail, ex);
```

Comment:
What happens when...

 - ex is null
 - ex is not a MessagingException
 - ex is `new MessagingException("not smtp related")`

?